### ISSUE TITLE
Add query to determine available bulk operations

### ISSUE THREAD

**Issue by eseliger** [sourcegraph/sourcegraph#21185]
Currently we do a best-effort in the frontend, but it would probably be simpler and more resilient to do that from the backend.

```graphql
availableBulkOperations(batchChange: ID!, changesets: [ID!]!): [BulkOperationType!]!
```

With such a GraphQL query we could change the dropdown in the frontend to not show an action when changeset(s) are selected but only once the dropdown button is hit. It would then query the available operation types and display those.

The components added in https://github.com/sourcegraph/sourcegraph/pull/22658 could maybe be reused here.

**Comment by malomarrec** [comment]
+1 encountered this again with customer. This is costing us support time, worth reconsidering the priority

**Comment by eseliger** [comment]
The problem we currently have with bulk operations is that certain preconditions for the changesets need to be met in order to run the bulk operation over them. For that, we have [logic in the frontend code that determines the available bulk operations](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/enterprise/batches/detail/changesets/ChangesetSelectRow.tsx?L48). It takes the filters from the changeset list, and if they meet the conditions, the bulk operation is available. That is very bad for discoverability though, because users need to KNOW what specific filter combination to enter to unlock a specific bulk operation. 
To solve that, we want to switch away from that conditional function in the frontend and rather ask the backend "hey the user selected changesets 1, 5 and 13, what bulk operations are available for those?". That's what the new query

```gql
availableBulkOperations(batchChange: ID!, changesets: [ID!]!): [BulkOperationType!]!
```

is supposed to serve. In the backend, we enforce specific conditions when actually running a bulk operation, which currently doesn't really match what the frontend did but we didn't have a better option for that.
They are [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1386-1390), [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1270-1273), [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1312-1317), [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1349-1352), [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1423-1427), and [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/batches/resolvers/resolver.go?L1459) (@LawnGnome is it correct that there are no preconditions for the publishChangesets bulk operation?).
That resolver method should do something like that:
- For all given changeset IDs, check what operations they currently support
- Increase a counter for each operation when a changeset supports it
- When the counter is len(changesetIDs), add it to the result list of operations
- Return the list of supported operations
(This is very unoptimized, there's hopefully a better way than this but logic wise I think it makes sense)

Once that's done, I think we can also remove [the check in the backend](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/internal/batches/service/service.go?L1120) that errors when a single changeset doesn't match the conditions anymore and simply skip it - the risk of it happening is low then and it reduces friction I guess.

**Comment by LawnGnome** [comment]
> @LawnGnome is it correct that there are no preconditions for the publishChangesets bulk operation?

Glad you asked!

Logically, since we don't support going `published -> draft`, we could exclude published, non-draft changesets from the bulk operation, and add a corresponding check. Practically, I chose not to implement a check because this wasn't going to map to the existing bulk operation UI (since I didn't want to tie the publish action to a particular set of filter states for discoverability reasons), and queuing a changeset job to publish an already published changeset is a (functional) no-op.

Here's a bonus I discovered after the fact: sometimes, when you're messing around with your local Sourcegraph reconciler, you end up leaving a changeset stuck in failed or errored. (Whichever one it is. I'm not looking it up right now.) Hitting publish actually forces the reconciler to run again, which is sometimes useful.

(Not saying we should keep it because of that, but I'd be lying if I said I hadn't used that trick several times in the last nine months.)
